### PR TITLE
feat: Add locale to route path

### PR DESCRIPTION
**Pull request by eugenio-d-tesio**
# 🤘 Flutter PRs

| :point_up: don't even dare to be here if you didn't check [before_pr.md](./docs/before_pr.md) file |
| --- |

Thank you for adding your efforts to this awesome project and Flutter Team! 🤘
Please, keep your PRs short to make everybody happy

_Made with love by the_ **Monkslab** ❤️

## Keep it clean

These are our selected best practices that we agree to follow for any PR to be approved. We think it's super handy for including everything without having to think about it. We hope they will make your life easier! 🙏

## Type

- [x] New feature
- [ ] Testing
- [ ] Refactor
- [ ] Bug-fix

## Versions

- [ ] upgrade with [Semantic Versioning](https://semver.org/) in pubspec.yaml: **latest here**

## Screenshots

[monkslab-web-28-02-23.webm](https://user-images.githubusercontent.com/5660624/228221039-d5b67996-993d-4c24-bc26-aff2a07f28b1.webm)

### CODE REVIEW

**Comment by maurodibert** [comment]
❤️!!!

File: lib/main.dart [hunk 1]
```diff
@@ -1,24 +1,26 @@
...
-        child: AppPage(),
+    MultiRepositoryProvider(
+      providers: [
+        RepositoryProvider(
+          create: (context) => coursesRepository,
+        ),
+      ],
+      child: BlocProvider(
+        create: (context) => LocaleCubit(),
+        child: const AppPage(),
```
Comment (EugenioTesio) [inline lib/main.dart:23]: Now it use a cubit to provide the locale

File: lib/src/app.dart [hunk 2]
```diff
@@ -1,54 +1,21 @@
...
-          locale: provider.locale,
-        );
-      },
+    return MaterialApp.router(
+      title: 'The MonksLab',
+      routeInformationParser: BeamerParser(),
+      routerDelegate: routerDelegate,
+      localizationsDelegates: AppLocalizations.localizationsDelegates,
+      supportedLocales: AppLocalizations.supportedLocales,
+      locale: context.watch<LocaleCubit>().state.locale,
```
Comment (EugenioTesio) [inline lib/src/app.dart:18]: Bloc watch to check locale changes

File: lib/src/app.dart [hunk 3]
```diff
@@ -1,54 +1,21 @@
...
-class LocaleProvider with ChangeNotifier {
-  Locale? _locale = const Locale('en');
-  Locale? get locale => _locale;
-
-  void setLocale(Locale locale) {
-    if (!AppLocalizations.supportedLocales.contains(locale)) return;
-    _locale = locale;
-    notifyListeners();
-  }
-}
```
Comment (EugenioTesio) [inline lib/src/app.dart:54]: Bye bye Provider

File: lib/src/core/routing/not_found_screen.dart [hunk 4]
```diff
@@ -7,9 +11,19 @@ class NotFoundScreen extends StatelessWidget {
...
+          children: [
+            Text('404 - Page not found!'.hardcoded),
+            AppGaps.gapH24,
+            TextButton(
+              onPressed: () => Beamer.of(context)
+                  .beamToReplacementNamed("/${context.read<LocaleCubit>().state.locale.languageCode}"),
+              child: const Text('Home'),
+            )
+          ],
+        ),
```
Comment (EugenioTesio) [inline lib/src/core/routing/not_found_screen.dart:26]: Removed the awful app bar and added a text button to redirect the user to home. If the language don¿t match with supported ones, will redirect to EN home

File: lib/src/core/routing/routing.dart [hunk 5]
```diff
@@ -8,22 +13,61 @@ enum AppRoutes {
...
+  locationBuilder: RoutesLocationBuilder(
+    routes: {
+      // Return either Widgets or BeamPages if more customization is needed
+      '/:locale': (_, state, ___) {
+        final languageCode = (state.pathParameters['locale']!);
+        return LocaleWrapper(
+          languageCode: languageCode,
+          child: const HomePage(),
+        );
+      },
```
Comment (EugenioTesio) [inline lib/src/core/routing/routing.dart:26]: In order to change locale with path, the pages needs to be wrapped with LocaleWrapper

File: lib/src/core/routing/routing.dart [hunk 6]
```diff
@@ -8,22 +13,61 @@ enum AppRoutes {
...
+  final Widget child;
+
+  @override
+  Widget build(BuildContext context) {
+    final locale = Locale(languageCode);
+    if (AppLocalizations.supportedLocales.contains(locale)) {
+      context.read<LocaleCubit>().setLocale(Locale(languageCode));
+      return child;
+    }
+    return const NotFoundScreen();
```
Comment (EugenioTesio) [inline lib/src/core/routing/routing.dart:71]: As said before, it the language isn't supported will return a NotFoundScreen

File: lib/main.dart (hunk 1 above)
Comment (maurodibert) [reply lib/main.dart:23]: Greeeat!

File: lib/src/app.dart (hunk 2 above)
Comment (maurodibert) [reply lib/src/app.dart:18]: Oh! This way whenever we switch the cubit, the whole thing changes?

File: lib/src/core/routing/not_found_screen.dart (hunk 4 above)
Comment (maurodibert) [reply lib/src/core/routing/not_found_screen.dart:26]: Niiice!

File: lib/src/core/routing/routing.dart (hunk 5 above)
Comment (maurodibert) [reply lib/src/core/routing/routing.dart:26]: I want a class of beamer! hehe

File: lib/src/core/routing/routing.dart (hunk 6 above)
Comment (maurodibert) [reply lib/src/core/routing/routing.dart:71]: Cooooool!

**Review by maurodibert — APPROVED** [review]
Tete! It works great and with lot of cool things over here! Awesome work! 💪

File: lib/src/core/routing/routing.dart (hunk 5 above)
Comment (EugenioTesio) [reply lib/src/core/routing/routing.dart:26]: Me too!!

File: lib/src/app.dart (hunk 2 above)
Comment (EugenioTesio) [reply lib/src/app.dart:18]: We, It rebuilds the whole Material app and under.

File: lib/src/core/routing/routing.dart (hunk 5 above)
Comment (maurodibert) [reply lib/src/core/routing/routing.dart:26]: hahahahah

File: lib/src/app.dart (hunk 2 above)
Comment (maurodibert) [reply lib/src/app.dart:18]: Cooo!

**Pull request merged**